### ISSUE TITLE
Debugger is very slow to start (21 seconds)

### ISSUE THREAD

**Issue by krassowski** [jupyterlab/jupyterlab#12012]
## Description

On a powerful desktop, enabling the debugger freezes the UI for way too long. This is a regression in 3.3 alpha.

## Reproduce

1. Install 3.3 in an environment with quite some number of packages installed (`pip list | wc -l` 337)
2. Open JupyterLab
3. Open a notebook with 195 code cells
4. Click on the bug icon
4. See freeze of about 20-30 seconds

## Expected behavior

The debugger starts within reasonable time without freezing UI, 2-3 seconds max.

## Context

- Operating System and version: Ubuntu
- Browser and version: Chrome 97
- JupyterLab version: 3.3.0a3

## Analysis

There are two issues here:

1. [ ] `set kernelSources` adds a huge number of widget nodes (~2 seconds); ideally we would make it 
2. [ ] [`_onCellsChanged`](https://github.com/jupyterlab/jupyterlab/blob/4dc0ec21fef81c18a912ea358800f5fd7e568dfa/packages/debugger/src/handlers/notebook.ts#L41) refreshes every single editor forcing hundreds of reflows via CodeMirror's `setOption` (~19 seconds)

Subsequently style recalculation takes additional 2-3 seconds due to big changes in the DOM.

![Untitled](https://user-images.githubusercontent.com/5832902/152522351-743aef56-0353-47da-b40b-de8793a7960f.png)

<details>

Zoom-in on `kernelSources`:

![Screenshot from 2022-02-04 11-44-40](https://user-images.githubusercontent.com/5832902/152523732-97514efe-6b6a-4c73-823e-66e7fe1523a7.png)

Zoom-in on `_onCellsChanged` in details:

![Screenshot from 2022-02-04 11-45-59](https://user-images.githubusercontent.com/5832902/152523867-7e4b86da-d46f-4d2c-bd47-9f14e1327616.png)

</details>

**Comment by echarles** [comment]
> There are two issues here:
> - set kernelSources adds a huge number of widget nodes (~2 seconds); ideally we would make it
> -  [_onCellsChanged](https://github.com/jupyterlab/jupyterlab/blob/4dc0ec21fef81c18a912ea358800f5fd7e568dfa/packages/debugger/src/handlers/notebook.ts#L41) refreshes every single editor forcing hundreds of reflows via CodeMirror's setOption (~19 seconds)

Do I understand that the new feature "Kernel Sources" counts for 2 seconds in 20-30 seconds freeze?

**Comment by krassowski** [comment]
Yes. The remaining 20-30 seconds is for refreshing line numbers/breakpoint marks.

**Comment by vidartf** [comment]
I cleaned up the marks in #12746. Can confirm that most of the remaining time is spent on kernel sources. It is easy to verify by simply adding a kernel source filter that narrows to only one file, and comparing that to the performance when the full list is present.

**Comment by MMesch** [comment]
The triaging agent has identified this issue as potentially resolved (rationale below). @krassowski can you confirm what the current status is and whether "The debugger starts within reasonable time without freezing UI, 2-3 seconds max."? If it is not the case, could you leave a comment making explicit what remains to be done if it is not the case?

---

**Agent Rationale**: The performance issues identified in this issue have been systematically addressed through multiple merged pull requests. Cross-referenced PRs #12746 and #12751 directly tackle the two main performance bottlenecks identified in the original issue. The current codebase shows evidence of the implemented optimizations including debounced filtering in kernel sources model and improved editor handling. Additionally, broader JupyterLab performance improvements in version 4.0+ (CodeMirror 6 upgrade, notebook windowing) further enhance debugger performance.
